[wip] prometheus monitoring support

snopoke opened the pull request:
This isn't done yet but putting up for initial review.

##### SUMMARY
Implementation of https://github.com/dimagi/commcare-hq/issues/26816

Initial code structure and classes to allow supporting multiple monitoring backends. Current support for Datadog and Prometheus.

I've converted a few metrics to see how it would work. Further notes in the code.

Missing pieces:
* the views to expose the metrics need to be added
* Prometheus initialisation
* Gunicorn hooks: https://github.com/prometheus/client_python#multiprocess-mode-gunicorn

Migration plan:
* this can be done piecemeal so once the basics are done I'll start with that
* there will likely be some decisions to make about changes to the metrics
  * split histograms out into their own metrics
  * should we adhere to a naming convention (e.g. https://prometheus.io/docs/practices/naming/)

snopoke committed:
move datadog init to AppConfig.ready

snopoke committed:
pluggable metrics

snopoke committed:
add histogram and tests

snopoke committed:
bucket values less than or equal

snopoke committed:
add prometheus client to requirements

snopoke committed:
Merge branch 'master' into sk/monitoring

snopoke committed:
make metrics lazy

snopoke committed:
example histogram

snopoke committed:
Merge branch 'master' into sk/monitoring

snopoke committed:
convert sumbission metrics

snopoke commented on corehq/apps/receiverwrapper/views.py:
```diff
@@ -172,24 +193,20 @@ def _submission_error(request, message, count_metric, metric_tags,
+        submission_duration_histogram.tag(**tags).observe(timer.duration)
 
-    datadog_counter('commcare.xform_submissions.count', tags=tags)
+    submission_counter.tag(**tags).inc()
```
This is an example of a metric where we have attached multiple histogram bucket tags to a single metric. If we want to convert these histograms to the native histogram types then they must be separated out as is done here.

snopoke commented on corehq/util/datadog/utils.py:
```diff
@@ -115,7 +115,7 @@ def bucket_value(value, buckets, unit=''):
     over_template = "over_{:0%s}{}" % number_length
     for bucket in buckets:
-        if value < bucket:
+        if value <= bucket:
```
Prometheus histograms uses `<=`. I changed this to make them compatible and to make testing easier but I can change if back. Note that I didn't change the tag name, it is still `lt` since that would break a bunch of DD graphs.

millerdev replied:
Once it is merged, this will change the meaning of metrics recorded in dd, right? That's not ideal, but I suppose in the long run it will not matter. It's also not great to have the tag name `lt` mean something different than the recorded values.

snopoke replied:
yes, that's correct. We could leave DD metrics as 'less than` and have Prometheus be `less than equal`, I would just need to update the tests to account for that.

snopoke replied:
Updated to keep this as `<` for DD

snopoke commented on corehq/util/metrics/__init__.py:
```diff
@@ -0,0 +1,15 @@
+    return enabled
+
+
+metrics = DelegatedMetrics(SimpleLazyObject(_get_metrics))  # singleton/global
```
Initially I had this as a function `get_metrics()` but after making it lazy there wasn't a need for the function. This will only call the `_get_metrics` function when a metric class is actually used. This is needed make sure that Django is fully set up before determining if the metrics providers are enabled (since they rely on settings).

```python
m = metrics.counter('name', 'description')  # still lazy
m.inc()  # now it's evaluated
```

millerdev replied:
Does this mean that `DelegatedMetrics` is always used even when there is only a single "metrics" object to delegate to? That seems wasteful. Would it be possible to only delegate if there is more than one (I would expect that to be the exceptional case).

snopoke replied:
Yes it does mean that. The alternative is to split `DelegatedMetrics` into two classes: `DelegatedMetrics` and `LazyMetrics` and then combine them as needed (the laziness is always needed).

snopoke commented on corehq/util/metrics/datadog.py:
```diff
@@ -0,0 +1,85 @@
+        _datadog_record(statsd.gauge, self.name, value, tags)
+
+
+class Histogram(HqHistogram):
```
As stated in the docstring this isn't a real histogram. It uses our workaround of adding bucketed tags.

snopoke commented on corehq/util/metrics/metrics.py:
```diff
@@ -0,0 +1,170 @@
+        "Did you mean to call your metric 'commcare.{}'? ".format(name))
+
+
+def _validate_tag_names(tag_names):
```
This is required by Prometheus.

snopoke commented on corehq/util/metrics/metrics.py:
```diff
@@ -0,0 +1,170 @@
+        self._record(value)
+
+
+DEFAULT_BUCKETS = (.005, .01, .025, .05, .075, .1, .25, .5, .75, 1.0, 2.5, 5.0, 7.5, 10.0, INF)
```
The default in Prometheus

snopoke committed:
docstrings

snopoke committed:
stickler

snopoke commented on corehq/apps/receiverwrapper/views.py:
```diff
@@ -50,18 +51,38 @@
 PROFILE_LIMIT = int(PROFILE_LIMIT) if PROFILE_LIMIT is not None else 1
 
+corrupt_multimedia_counter = metrics.counter(
+    'commcare.corrupt_multimedia_submissions', 'Count of requests with corrupt multimedia',
```
This name is changed
```diff
- commcare.corrupt-multimedia-submission.error.count
+ commcare.corrupt_multimedia_submissions
```

I changed this because `.error` seemed redundant since it's implied by the fact that it's 'corrupt'. I dropped the 'count' suffix because Prometheus adds a `_total` prefix to the end of counter metric names. If we want to keep the suffix we should add it to all 'counter' metrics at the provider level.

millerdev commented on corehq/util/metrics/metrics.py:
```diff
@@ -47,38 +53,54 @@ def tag(self, **tag_kwargs):
+        self._bucket_unit = self._kwargs.get('bucket_unit', '')
+        self._bucket_tag = self._kwargs.get('bucket_tag')
+        if self._bucket_tag in self.tag_names:
+            self.tag_names = tuple([name for name in self.tag_names if name != self._bucket_tag])
```
nit
```suggestion
            self.tag_names = tuple(name for name in self.tag_names if name != self._bucket_tag)
```

millerdev commented on corehq/util/metrics/metrics.py:
```diff
@@ -30,14 +31,19 @@ def _validate_tag_names(tag_names):
         self.tag_names = _validate_tag_names(tag_names)
-        self.tag_values = tag_values
+        self.tag_values = kwargs.pop('tag_values', None)
+        self._kwargs = kwargs
```
Will an error be raised if unknown kwargs (due to typo, etc.) are passed in here? It looks like the answer is "No." Would be great if that situation could be improved to make it more typo-resistant.

snopoke replied:
I agree, this was another part I wasn't very happy with. I struggled to find a clean way to do this. It stems from the fact that the Datadog and Prometheus classes require different arguments. One good thing here is that these classes are never instantiated outside of the package. The public API has all the available kwargs clearly listed.

I think this stems from trying to emulate the Prometheus API: `metric.label(...).inc()` which requires classes.

millerdev commented on corehq/util/metrics/tests/test_metrics.py:
```diff
@@ -0,0 +1,171 @@
+class TestDatadogMetrics(_TestMetrics):
+    provider_class = DatadogMetrics
+
+    def setUp(self) -> None:
```
Do type annotations like this do anything useful?

snopoke replied:
No, it should be removed. This was added by autocomplete in pycharm and I never removed it.

millerdev commented on corehq/util/metrics/metrics.py:
```diff
@@ -0,0 +1,172 @@
+    def enabled(self) -> bool:
+        raise NotImplementedError
+
+    def counter(self, name: str, documentation: str, tag_names: Iterable = tuple()) -> HqCounter:
```
```suggestion
    def counter(self, name: str, documentation: str, tag_names: Iterable = ()) -> HqCounter:
```
This suggestion can also be applied in several other places throughout this PR, but I have only marked this one.

snopoke replied:
0bcedea6a39

millerdev commented on corehq/util/metrics/tests/utils.py:
```diff
@@ -0,0 +1,22 @@
+    def record(fn, name, value, tags=None):
+        key = tuple([
+            name,
+            tuple([tuple(t.split(':')) for t in (tags or [])]),
```
nit
```suggestion
            tuple(tuple(t.split(':')) for t in (tags or [])),
```

snopoke replied:
450f211179d

millerdev commented on corehq/util/metrics/tests/utils.py:
```diff
@@ -0,0 +1,22 @@
+    try:
+        yield stats
+    finally:
+        patch.stop()
```
```suggestion
    with mock.patch("corehq.util.metrics.datadog._datadog_record", new=record):
        yield stats
```

snopoke replied:
Copied from here https://github.com/dimagi/commcare-hq/blob/ebd9d113c0ca9a1bdddcdb2c1fda8a84180404ad/corehq/util/test_utils.py#L591-L608. Was planning to remove the old one once migration is complete.

Updated the new one: 450f211

millerdev replied:
ha! I don't know why I originally wrote it that way. 👍 on eventual removal.

millerdev reviewed (commented):
How was Prometheus chosen? Is it the clear leader in the open source monitoring space, or are there other options?

About the naming convention, it looks like all of our current metric names violate the Prometheus default since they use dot (`.`) as a separator (they also use underscore (`_`) as sub-separator), where Prometheus uses underscore only. Do I have that right?

snopoke commented on corehq/util/metrics/metrics.py:
```diff
@@ -0,0 +1,139 @@
+
+METRIC_NAME_RE = re.compile(r'^[a-zA-Z_:.][a-zA-Z0-9_:.]*$')
+METRIC_TAG_NAME_RE = re.compile(r'^[a-zA-Z_][a-zA-Z0-9_]*$')
+RESERVED_METRIC_TAG_NAME_RE = re.compile(r'^__.*$')
```
True, this was copied from the Prometheus client.

snopoke committed:
keep tag_values as dict instead of splitting and re-combining

snopoke commented on corehq/util/metrics/metrics.py:
```diff
@@ -0,0 +1,139 @@
+        self.tag_values = tag_values
+
+    def tag(self, **tag_kwargs):
+        if sorted(tag_kwargs) != sorted(self.tag_names):
```
450f211179d

snopoke commented on corehq/util/metrics/prometheus.py:
```diff
@@ -0,0 +1,46 @@
+        return self.__class__(self.name, self.documentation, self.tag_names, self.tag_values, delegate)
+
+
+class Counter(PrometheusMetricBase, HqCounter):
```
Yes, the inheritance isn't great here. The alternative is a bit more duplicate code but not a lot. I'll see what I can do to make it better if we stay on this course.

If we want to keep it functional then we would need to have a centralized metrics singleton that would always return the same metric class when called (in the case of Pometheus). It would also mean that every call to the metric would need to pass in all required args (which is the case now anyway):
```python
metrics_histogram(value, tags=..., buckets=..., unit=..., dd_bucket_tag=...)
```

snopoke replied:
I've resolved this by removing the need to create new copies of the class when applying tags:
```diff
- metric.tag(group='a').inc()
+ metric.inc(group='a')
```

snopoke committed:
update links

snopoke commented on corehq/util/metrics/datadog.py:
```diff
@@ -30,20 +36,48 @@ def _datadog_record(fn, name, value, tags=None):
+        # resulting Datadog metric
+        #    commcare.request.duration:1|c|#duration:lt_2ms
+
+    More details: https://help.datadoghq.com/hc/en-us/articles/211545826
```
This link was also copied from old DD code. I wasn't aware it redirected. I'll replace it with the other links: 2c3edefb732

snopoke committed:
remove unnecessary list

Co-Authored-By: Daniel Miller <dmiller@dimagi.com>

snopoke committed:
replace typle() with ()

snopoke committed:
Merge branch 'sk/monitoring' of github.com:dimagi/commcare-hq into sk/monitoring

snopoke commented:
> How was Prometheus chosen? Is it the clear leader in the open source monitoring space, or are there other options?

We did a basic survey. There a few options but Prometheus appears to be the more common one and support all the use cases we need.

> About the naming convention, it looks like all of our current metric names violate the Prometheus default since they use dot (.) as a separator (they also use underscore (_) as sub-separator), where Prometheus uses underscore only. Do I have that right?

That's correct. I've already taken that into account by replacing `.` with `_` for Prometheus so I'm not suggesting we change that. I was meaning more of the semantic guidelines.

snopoke committed:
fix tags

snopoke committed:
pass other args

snopoke committed:
revert change to datadog bucketing boundry

snopoke commented:
Note to reviewers: I'm offline till Monday. I'd like to schedule a call for when I'm back to decide on a course of action for this. If you'd like to weigh in please review by then (9 am EST on 16 March)

millerdev commented on corehq/util/metrics/tests/utils.py:
```diff
@@ -9,14 +9,10 @@ def patch_datadog():
         key = tuple([
             name,
-            tuple([tuple(t.split(':')) for t in (tags or [])]),
+            tuple(sorted([tuple(t.split(':')) for t in (tags or [])])),
```
nit... lol
```suggestion
            tuple(sorted(tuple(t.split(':')) for t in (tags or []))),
```

snopoke committed:
remove unnecessary list

snopoke committed:
apply tags at the same time as recording the metric

This change means we don't have to instantiate a new
class when applying the tags which simplifies the class
structure a lot.

snopoke committed:
dummy metric

snopoke commented:
closing in favour of https://github.com/dimagi/commcare-hq/pull/26876
